src/shared/mount-util.c

```diff
@@ -490,52 +490,81 @@ int mount_move_root(const char *path) {
+                prop_name = "private";
+                break;
+        case PROPAGATION_DEPENDENT:
+                flags = MS_SLAVE;
+                prop_name = "dependent";
+                break;
+        case PROPAGATION_SHARED:
+                flags = MS_SHARED;
+                prop_name = "shared";
+                break;
+        default:
+                assert_not_reached();
+        }
 
-        /* Get rid of the old root and reveal our brand new root. */
-        if (umount2(".", MNT_DETACH) < 0)
-                return log_debug_errno(errno, "Failed to unmount old rootfs: %m");
+        if (in_initrd()) {
+                int r;
+                /* pivot_root() isn't currently supported in the initramfs so fallback to MS_MOVE. */
+                r = mount_move_root(path);
+                if (r < 0)
+                        return log_debug_errno(r, "Failed to move new rootfs '%s': %m", path);
+        } else {
+                fd_oldroot = open("/", O_PATH | O_DIRECTORY | O_CLOEXEC | O_NOFOLLOW);
+                if (fd_oldroot < 0)
+                        return log_debug_errno(errno, "Failed to open old rootfs");
+
+                fd_newroot = open(path, O_PATH | O_DIRECTORY | O_CLOEXEC | O_NOFOLLOW);
+                if (fd_newroot < 0)
+                        return log_debug_errno(errno, "Failed to open new rootfs '%s': %m", path);
+
+                /* Change into the new rootfs. */
+                if (fchdir(fd_newroot) < 0)
+                        return log_debug_errno(errno, "Failed to change into new rootfs '%s': %m", path);
+
+                /* Let the kernel tuck the new root under the old one. */
+                if (pivot_root(".", ".") < 0)
+                        return log_debug_errno(errno, "Failed to pivot root to new rootfs '%s': %m", path);
+
+                /* At this point the new root is tucked under the old root. If we want
+                 * to unmount it we cannot be fchdir()ed into it. So escape back to the
+                 * old root. */
+                if (fchdir(fd_oldroot) < 0)
+                        return log_debug_errno(errno, "Failed to change back to old rootfs: %m");
+
+                /* Note, usually we should set mount propagation up here but we'll
+                 * assume that the caller has already done that. */
+
+                /* Get rid of the old root and reveal our brand new root. */
+                if (umount2(".", MNT_DETACH) < 0)
+                        return log_debug_errno(errno, "Failed to unmount old rootfs: %m");
+
+                if (fchdir(fd_newroot) < 0)
+                        return log_debug_errno(errno, "Failed to switch to new rootfs '%s': %m", path);
+        }
 
-        if (fchdir(fd_newroot) < 0)
-                return log_debug_errno(errno, "Failed to switch to new rootfs '%s': %m", path);
+        /* Finally, let's establish the requested propagation type. */
+        if ((flags != 0) && mount(NULL, "/", NULL, flags | MS_REC, 0) < 0)
```

Comment:
`()` is redundant.

Reply:
I know but I like the semantic clarity of having them around instead of just relying on operator precedence.